### PR TITLE
Add decay to CyclicLrUpdater

### PR DESCRIPTION
**Pull request by HAOCHENYE**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

I followed #1059 to modify and fixed the cyclic lr decay.

## Modification

Fixed lr update error and unit test. The lr curve of config:
```
lr_config = dict(
    _delete_="step",
    policy='cyclic',
    # anneal_strategy='linear',
    target_ratio=(10, 1),
    cyclic_times=5,
    step_ratio_up=0.4,
    gamma=0.5,
)
```
image
```
lr_config = dict(
    _delete_="step",
    policy='cyclic',
    # anneal_strategy='linear',
    target_ratio=(10, 1e-4),
    cyclic_times=5,
    step_ratio_up=0.4,
    gamma=0.5,
)
```
image
```
lr_config = dict(
    _delete_="step",
    policy='cyclic',
    # anneal_strategy='linear',
    target_ratio=(0.5, 1),
    cyclic_times=5,
    step_ratio_up=0.4,
    gamma=0.5,
) # support target_ratio[0] < 1
```
image

## BC-breaking (Optional)

No.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here, and update the documentation.

## Checklist

**Before PR**:

- [x] I have read and followed the workflow indicated in the [CONTRIBUTING.md](https://github.com/open-mmlab/mmcv/blob/master/CONTRIBUTING.md) to create this PR.
- [x] Pre-commit or linting tools indicated in [CONTRIBUTING.md](https://github.com/open-mmlab/mmcv/blob/master/CONTRIBUTING.md) are used to fix the potential lint issues.
- [x] Bug fixes are covered by unit tests, the case that causes the bug should be added in the unit tests.
- [ ] New functionalities are covered by complete unit tests. If not, please add more unit test to ensure the correctness.
- [ ] The documentation has been modified accordingly, including docstring or example tutorials.

**After PR**:

- [ ] If the modification has potential influence on downstream or other related projects, this PR should be tested with some of those projects, like MMDet or MMCls.
- [ ] CLA has been signed and all committers have signed the CLA in this PR.

### CODE REVIEW

**Review by teamwong111 — APPROVED** [review]
LGTM

File: mmcv/runner/hooks/lr_updater.py [hunk 1]
```diff
@@ -420,15 +420,16 @@ class CyclicLrUpdaterHook(LrUpdaterHook):
...
+        by_epoch (bool, optional): Whether to update LR by epoch.
+        target_ratio (tuple[float], optional): Relative ratio of the highest LR
+            and the lowest LR to the initial LR.
+        cyclic_times (int, optional): Number of cycles during training
+        step_ratio_up (float, optional): The ratio of the increasing process of
+            LR in the total cycle.
+        anneal_strategy (str, optional): {'cos', 'linear'}
             Specifies the annealing strategy: 'cos' for cosine annealing,
             'linear' for linear annealing. Default: 'cos'.
+        gamma (float, optional): Cycle decay ratio. Default: 1.
```
Comment (zhouzaida) [inline mmcv/runner/hooks/lr_updater.py:432]:
```suggestion
        gamma (float, optional): Cycle decay ratio. Default: 1.
            `New in version 1.4.4.`
```

**Comment by zhouzaida** [comment]
It would be nice if an image can be provided in the PR message to visualize the learning rate. Maybe you can refer to https://mmclassification.readthedocs.io/en/latest/tools/visualization.html#learning-rate-schedule-visualization.

File: mmcv/runner/hooks/lr_updater.py [hunk 2]
```diff
@@ -473,21 +482,27 @@ def before_run(self, runner):
...
+        # Update weight decay
+        scale = self.gamma**curr_cycle
+
+        for (start_iter, end_iter, start_ratio, end_ratio) in self.lr_phases:
             if start_iter <= curr_iter < end_iter:
+                # Apply cycle scaling to gradually reduce max_lr.
+                if end_ratio > start_ratio:
+                    end_ratio = 1 - scale + end_ratio * scale
+                else:
+                    start_ratio = 1 - scale + start_ratio * scale
```
Comment (zhouzaida) [inline mmcv/runner/hooks/lr_updater.py:505]: We need to add a comment to describe why the end_ratio will be `1 - scale + end_ratio * scale` like `To ensure that peak_lr is larger than base_lr, we add 1-scale as a bias. If self.target_ratio is greater than 1, this ensures that end_ratio or start_ratio is always greater than 1, i.e. peak_lr is greater than base_lr.` In addition, we should add more description in the pr message with several figures.

**Comment by zhouzaida** [comment]
We also need to add the two arguments `anneal_strategy` and `gamma` for `CyclicMomentumUpdaterHook`.
https://github.com/open-mmlab/mmcv/blob/86ed509a8bc783cdd0617efeba257375d5aa6658/mmcv/runner/hooks/momentum_updater.py#L226

File: mmcv/runner/hooks/momentum_updater.py [hunk 3]
```diff
@@ -276,23 +294,35 @@ def before_run(self, runner):
...
+        curr_iter = runner.iter % self.max_iter_per_phase
+        curr_cycle = runner.iter // self.max_iter_per_phase
+        scale = self.gamma**curr_cycle
+        for (start_iter, end_iter, start_ratio, end_ratio) \
+                in self.momentum_phases:
             if start_iter <= curr_iter < end_iter:
+                # Apply cycle scaling to gradually reduce the difference
+                # between max_lr and base lr. The target end_ratio can be
+                # expressed as:
+                # end_ratio = (base_lr + scale * (max_lr - base_lr)) / base_lr
```
Comment (zhouzaida) [inline mmcv/runner/hooks/momentum_updater.py:315]: lr->momentum

**Review by zhouzaida — APPROVED** [review]
The PR can be merged after resolving the comment.

**Review by zhouzaida — APPROVED** [review]

**Review by ZwwWayne — APPROVED** [review]

**Pull request merged**